src/pcap/api.go

```diff
@@ -0,0 +1,650 @@
+func readMsgFromStream(ctx context.Context, captureStream captureReceiver, target AgentEndpoint, bufSize int) <-chan *CaptureResponse {
+	out := make(chan *CaptureResponse, bufSize)
+	stopped := false
+	go func() {
+		defer close(out)
+		// defer wg.Done()
+		defer func(captureStream captureReceiver) {
+			closeSendErr := captureStream.CloseSend()
+			if closeSendErr != nil {
+				out <- convertStatusCodeToMsg(closeSendErr, target)
+				return
+			}
+		}(captureStream)
+		for {
+			if ctx.Err() != nil && !stopped {
+				stopped = true
+				err := captureStream.Send(&AgentRequest{
+					Payload: &AgentRequest_Stop{},
+				})
+				if err != nil {
+					out <- convertStatusCodeToMsg(err, target)
+					return
+				}
+			}
+			msg, err := captureStream.Recv()
+			if err != nil && errors.Is(err, io.EOF) {
+				msg := fmt.Sprintf("Capturing stopped on agent pcap-agent %s", target)
+				out <- newMessageResponse(MessageType_CAPTURE_STOPPED, msg, target.Identifier)
+				return
+			}
+			code := status.Code(err)
+			if code != codes.OK {
+				out <- convertStatusCodeToMsg(err, target)
+				return
+			}
+			out <- msg
+		}
+	}()
+	return out
+}
+
+func convertStatusCodeToMsg(err error, target AgentEndpoint) *CaptureResponse {
+	code := status.Code(err)
+	if code == codes.Unknown {
+		unwrappedError := errors.Unwrap(err)
+		if unwrappedError != nil {
+			code = status.Code(unwrappedError)
+		}
+	}
+	err = fmt.Errorf("capturing from agent %s: %w", target, err)
+
+	//FIXME: internal+unknown and default are the same. Is default really a connection error?
+	switch code { //nolint:exhaustive // we do not need to cover all the codes here
+	case codes.InvalidArgument:
+		return newMessageResponse(MessageType_INVALID_REQUEST, err.Error(), target.Identifier)
+	case codes.Aborted:
+		return newMessageResponse(MessageType_INSTANCE_UNAVAILABLE, err.Error(), target.Identifier)
+	case codes.Internal, codes.Unknown:
+		return newMessageResponse(MessageType_CONNECTION_ERROR, err.Error(), target.Identifier)
+	case codes.FailedPrecondition:
+		return newMessageResponse(MessageType_START_CAPTURE_FAILED, err.Error(), target.Identifier)
+	case codes.ResourceExhausted:
+		return newMessageResponse(MessageType_LIMIT_REACHED, err.Error(), target.Identifier)
+	case codes.Unavailable:
+		return newMessageResponse(MessageType_INSTANCE_UNAVAILABLE, err.Error(), target.Identifier)
+	default:
+		return newMessageResponse(MessageType_CONNECTION_ERROR, err.Error(), target.Identifier)
+	}
+}
+
+// boshRequestReceiver is an interface used by boshStopCmd to simplify testing.
+type requestReceiver interface {
+	Recv() (*CaptureRequest, error)
+}
+
+// stopCmd reads the next message from the stream. It ensures that the message
+// has a payload of StopBoshCapture. If any error is encountered or the payload is
+// of a different type an appropriate cause is set and the cancel function is called.
+func stopCmd(cancel CancelCauseFunc, stream requestReceiver) {
```

Comment:
Duplicate of `agentStopCmd`. Again a case for refactoring those two functions to one to serve a similar purpose. Also the implementation differ, not sure why but this should be aligned.

Reply:
It is not really a duplicate as we use different requests structs for api ([CaptureRequest](https://github.com/cloudfoundry/pcap-release/blob/main/src/pcap/pcap.proto#L89-L94)) and for agent ([AgentRequest](https://github.com/cloudfoundry/pcap-release/blob/main/src/pcap/pcap.proto#L138-L144)) with different properties. I am not sure if we want to extract stopRequest to separate struct independent on request type. 
@maxmoehl @peanball WDYT?